### PR TITLE
Use C++ concepts for propagate_const_array

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Switch to use concepts to constrain which types the class supports.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Utilities/interface/propagate_const_array.h [hunk 1]
```diff
@@ -19,18 +19,9 @@
 
 namespace edm {
 
+  template <typename T>
+  concept ArrayAddressable = requires(T& a) { a[0]; } or std::is_array_v<T>;
```
Comment (makortel) [inline FWCore/Utilities/interface/propagate_const_array.h:23]: Do we want this `ArrayAddressable` to be potentially used elsewhere, or is it more of an implementation detail of `propagate_const_array`?

Comment (Dr15Jones) [reply FWCore/Utilities/interface/propagate_const_array.h:23]: I was of two minds for that. I THINK it should only be here. If it proves useful, we could later move it.

Comment (makortel) [reply FWCore/Utilities/interface/propagate_const_array.h:23]: If we treat it as an implementation detail, should it be placed in `edm::impl` namespace?

Comment (Dr15Jones) [reply FWCore/Utilities/interface/propagate_const_array.h:23]: I thought about that but really, `edm::` and `edm::impl` are really not that much different for _polluting namespace_ reasons. If we REALLY wanted to be better than we should use something like `edm::propagate_impl` to better isolate names. 

I'm not really arguing against using `edm::impl`, it's just that we use that a lot.

Comment (makortel) [reply FWCore/Utilities/interface/propagate_const_array.h:23]: I was thinking more from the "hiding from users" point of view than (along `edm::impl` _hints_ that users shouldn't use it). I'd be perfectly fine with `edm::propagate_impl` here (and I agree with the name collision prevention argument).

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
I found a bug in the requires clause

```C++
 template <typename T>
      requires requires {
        not std::is_array_v<T>;
        requires ArrayAddressable<T>;
      }
    struct subscript_type<T> 
```

should be
```C++
 template <typename T>
      requires requires {
        requires not std::is_array_v<T>;
        requires ArrayAddressable<T>;
      }
    struct subscript_type<T> 
```

the previous just says the expression `not std::is_array_v<T>` has to compile, while `requires not std::is_array_v<T>` says the value must evaluate to `true`

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+core

**Comment by makortel** [comment]
ignore tests-rejected with ib-failure

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**